proposals/0012-introduce-reactNativeMeta.md

```diff
@@ -0,0 +1,148 @@
+
+It became clear that there was a need to be able to communicate and parse certain information about a react-native library via its `package.json`, which led to [a meeting](https://github.com/microsoft/rnx-kit/discussions/2125) between Meta, Microsoft and Expo to figure out a common solution.
+
+This RFC is the result of that meeting: a new section of the `package.json`, called `reactNativeMeta`, to contain certain metadata about a react-native project - it being either a library or an app.
+
+There are a few areas that this section could express in a consistent manner; here's a list of the ones we could think of (in no particular order):
+
+- if a library supports the new architecture, turbo modules or fabric or both
+- which versions of react-native a library supports in a given release
+- if an app has opted in into Fabric, TM, or both
+- if an app is using Hermes or another js engine
+
+Sidenote: the name `reactNativeMeta` was chosen to mirror the camel case pattern used by other fields such as `peerDependenciesMeta`.
+
+## Detailed design
+
+This is the section that in this first draft will need to be properly finalised. For now, here's what we envisioned:
+
+```js
+reactNativeMeta: {
+    version: string, // ex. 1.0.0 - we know that this spec might evolve in the future
+    type: string; // "app" or "library"
+    features: {
+        // can be global...
+        newArchEnabled: boolean,
+        jsEngine: string, // ex. "hermes", "jsc", "V8"...
+        // ...or platform specific
+        android: {
+            newArchEnabled: boolean,
+            fabric: boolean,
+            turbomodules: boolean,
+            jsEngine: string,
+        },
+        iOs: {
+        ... // same as android
+        },
+        macOs: {
+        ... // same as android
+        },
+        windows: {
+        ... // same as android
+        },
+        // more platforms, params - add a comment here
+    },
+    // this section would most likely only be relevant for libraries
+    requirements: {
+        "react-native": string, // semver compliant range of versions of RN supported
+        expo: string, // semver compliant range of versions of Expo supported (if any)
+    }
+}
+```
+
+## Drawbacks
+
+The main drawback is that we'd need to convince developers to start using it, so we'd have to expect a phase in which only a part of the userbase has this section configured and tool developers need to account for the section missing entirely.
+
+## Alternatives
+
+`reactNativeMeta` is, by design, fully original and new: there are other files that are used to define configurations for various aspect of react-native based projects (such as `react-native.config.js`, `app.json`, Expo's `app.config.js`, `expo-module.config.json`) but this one does **not** overlap with any of them. Explicitly, this one has the purpose of being filled only with metadata to be read by package managers and other tools.
```

Comment:
As this grows, or if we want to allow it to be dynamic, or extendable, it seems beneficial to allow mirroring into a `react-native.config.js` outside the direct package.json (maybe not that since already taken).  E.g. see recent template change moving Jest config out of package.json (or... cosmic-config bits).

Reply:
`react-native.config.js` already exists and is used for CLI stuff (see https://github.com/react-native-community/cli/blob/main/docs/configuration.md#configuration) - it is an explicit goal of this metadata field to be entirely separate.

Reply:
Call it `react-native-meta.config.js` as an example. But it seems like the app fields are being used for configuration. If that configuration becomes non-trivial, it may scale badly inlined into each package.json vs being shareable. The mapping of `foo` in package.json to `foo.config.js` is also somewhat expected by convention imo.

Reply:
> filled only with metadata to be read by package managers and other tools

What does this mean? It seems like it allows it to be used for almost anything.

Reply:
> As this grows, or if we want to allow it to be dynamic, or extendable from a base, it seems beneficial to allow mirroring into a react-native.config.js outside the direct package.json 

Could you clarify on why we want this now? This is something we can still add later if we wish and the current design is not preventing it. Or am I missing something?